### DISCUSSION TITLE
how to parse payload of raw transaction

### DISCUSSION THREAD

**Discussion by eden-wang** [starcoinorg/starcoin#2457] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/25053532/116334931-9d77a200-a808-11eb-8670-b5303f9ced4a.png)
![image](https://user-images.githubusercontent.com/25053532/116334966-abc5be00-a808-11eb-9803-39e4da03d4a8.png)

I want to get recipient's address and amount, but I don't know how to parse the payload. I do know what I need is in the payload

**Comment by jolestar** [comment, marked as answer]
The document is out of date, the payload struct is defined at https://github.com/starcoinorg/starcoin/blob/99f03927f5280b3c494a0d4a92ad4c94a859466d/vm/types/src/transaction/mod.rs#L305

What is the language that you used to parse the payload?

**Reply by eden-wang** [reply]
I use python. I found the [python sdk](https://github.com/starcoinorg/starcoin-sdk-python) yesterday, and it works now, thank you
```
    from starcoin.starcoin_types import TransactionPayload__ScriptFunction
    p = TransactionPayload__ScriptFunction.bcs_deserialize(
        bytes.fromhex(
            "02000000000000000000000000000000010f5472616e73666572536372697074730c706565725f746f5f706565720107000000000000000000000000000000010353544303535443000310f5ae543d68bfa9ff62f2f9b26f6f058d01001000e1f505000000000000000000000000")
    )
    r = p.value.args[0].hex()
    amount = p.value.args[2]
    from starcoin import bcs
    amount = bcs.deserialize(amount, st.uint128)[0]
    print(p)
    print(r, int(amount) * 10 ** -9)
```
the result is
```
TransactionPayload__ScriptFunction(value=ScriptFunction(module=ModuleId(address=AccountAddress(value=(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1)), name=Identifier(value='TransferScripts')), function=Identifier(value='peer_to_peer'), ty_args=[TypeTag__Struct(value=StructTag(address=AccountAddress(value=(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1)), module=Identifier(value='STC'), name=Identifier(value='STC'), type_params=[]))], args=[b'\xf5\xaeT=h\xbf\xa9\xffb\xf2\xf9\xb2oo\x05\x8d', b'\x00', b'\x00\xe1\xf5\x05\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00']))
f5ae543d68bfa9ff62f2f9b26f6f058d 0.1
```

**Comment by tarvos21** [comment]
[starcoin.js](https://github.com/starcoinorg/starcoin.js/blob/0688f8f2d804e4bb7da0f4bcd5b08431b0f60a14/src/encoding/index.spec.ts#L39) also provides function to decode transaction payload. Examples can be found here:

- [decode transaction payload](https://github.com/starcoinorg/starcoin.js/blob/0688f8f2d804e4bb7da0f4bcd5b08431b0f60a14/src/encoding/index.spec.ts#L39) 

- [starcoin-explorer's transaction details](https://github.com/starcoinorg/starcoin-explorer/blob/8b5dbb77b803d2cabc4906d61de9c5ab0abf67ec/src/modules/Transactions/components/Detail/index.tsx#L100)

**Reply by eden-wang** [reply]
yeah, it is. But python sdk works for me right now, thank you
